### DISCUSSION TITLE
[xstate/react] Memoized guards using useMachine

### DISCUSSION THREAD

**Discussion by bartekczyz** [statelyai/xstate#1979] [Q&A]
First of all, thank you for this awesome lib! 🤝 

**Description**

I need to use a guard with a dynamic data. Instead of defining it in the machine, I do it through `options` argument of `useMachine` hook. It works partially, because it memoizes the guard function. It works just fine with the `actions` object, but it fails with `guards`. I've tried patching [the hook](https://github.com/davidkpiano/xstate/blob/master/packages/xstate-react/src/useInterpret.ts#L125-L127) with following code:

```tsx
useIsomorphicLayoutEffect(() => {
  Object.assign(service.machine.options.actions, actions);
  Object.assign(service.machine.options.guards, guards);
}, [actions, guards]);
```

but unfortunately it didn't work.

**Expected Result**

I should be able to define a guard that can access component's latest props

**Actual Result**

Once the component is rendered, all guards are memoized

**Reproduction**

- use [this sandbox](https://codesandbox.io/s/admiring-bhabha-lopps)
- open console
- click `Regenerate random value` - you should see the value in the console
- click `Exec` - it would print initial "randomValue"

**Additional context**

**Use case**: I'm building a `DateRangePicker` component. It has `minDate` and `maxDate` props, some of the guards rely on these constraints. Both `minDate` and `maxDate` can be changed dynamically in a runtime, so visually it looks just fine, but the State Machine "remembers" the values from the 1st render. As a workaround, we pass a custom key, but this is annoying and component consuming developer shouldn't worry about that. Here's the workaround:
```tsx
<DateRangePicker 
  key={ /* here we had to figure out a good key */ } 
  minDate={minDate} 
  maxDate={maxDate} 
  value={value} 
  onChange={handleChange} 
/>
```

**XState version**: latest

**Comment by davidkpiano** [comment]
<sup>(Converted to a discussion because this is more of a question rather than an issue)</sup>

With state machines, any changes must be explicit. Also, guards typically shouldn't rely on outside information; they should be **pure functions**.

One solution is to create events/transitions for `"minMaxDate.update"` (or similar) and send those events when those props update:

```js
useEffect(() => {
  send('minMaxDate.update', { minDate, maxDate });
}, [minDate, maxDate]);

// ...
on: {
  'minMaxDate.update': {
    actions: assign({ minDate: (_, e) => e.minDate, maxDate: (_, e) => e.maxDate })
  }
}
// ...
```

Then your guards don't have to be dynamic. Even better, you now have a machine that is more testable, more robust (since your guard functions are pure and there is no implicit state), and you can achieve what you want in fewer lines of code than the `Object.assign()` mutations you're doing above.

**Reply by bartekczyz** [reply]
Thank you for this suggestion!
